### PR TITLE
ENH: add db_utils and the key hunting thing

### PR DESCRIPTION
**Pull request by CJ-Wright**

### CODE REVIEW

**Comment by CJ-Wright** [comment]
@sbillinge ready for review. (also I need a name for one of the functions)

File: xpdan/db_utils.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
+from pprint import pprint
+from .utils import _timestampstr
+
+
+def need_name_here(hdrs, key, verbose=True):
```
Comment (sbillinge) [inline xpdan/db_utils.py]: sort_scans_by_hdr_key()

File: xpdan/db_utils.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+                d[hdr.start[key]].append(i)
+            else:
+                d[hdr.start[key]] = [i]
+    if verbose:
+        pprint(d)
+    return d
+
+
+def scan_diff(hdrs, verbose=True):
+    """Get the metadata differences between scans
```
Comment (sbillinge) [inline xpdan/db_utils.py]: add '...between scans in hdrs list'

This function prints all the header information from each header in the hdrs list, except for any key:value pairs that have the same value for all the headers in the list.

File: xpdan/db_utils.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+def scan_headlines(hdrs, fields=None, verbose=True):
+    """Provide one line summaries of headers
+
+    Parameters
+    ----------
+    hdrs: list of headers
+        The headers from the databroker
+    fields: list of str, optional
+        The fields to be included in the summary, if None use
```
Comment (sbillinge) [inline xpdan/db_utils.py]: 'Specify a list of fields to summarize.  If None, the following will be returned'

File: xpdan/db_utils.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    verbose: bool, optional
+        If True print the summary
+
+    Returns
+    -------
+    list:
+        List of summary strings
+    """
+    if fields is None:
+        fields = ['sample_name', 'temperature', 'diff_x', 'diff_y', 'eurotherm']
```
Comment (CJ-Wright) [inline xpdan/db_utils.py]: scan plan type

Comment (CJ-Wright) [inline xpdan/db_utils.py]: Also the temperature start and end

Comment (sbillinge) [inline xpdan/db_utils.py]: 'sample_name', 'temperature', 'diff_x', 'diff_y', 'eurotherm','cs700' (not sure, but we want the temperature returned by the cryostream controller), count time, scan_type, start_temp, end_temp, temp_step, exposure, pause

note to self: make an issue sot hat this is controlled by global/config later

File: xpdan/db_utils.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
+from pprint import pprint
+from .utils import _timestampstr
+
+
+def need_name_here(hdrs, key, verbose=True):
+    """Build a dictionary where the keys are the unique values for the given
```
Comment (sbillinge) [inline xpdan/db_utils.py]: In a list of hdrs, group the scans by header-key.

Use this function to find all the scans in the list of headers that have a particular key value, such as 'sample_name'='Ni'

function returns a list of indices for the position in the list, associated with the metadata key values.

File: xpdan/db_utils.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    key and the values are the positions of the headers for that key
+
+    Parameters
+    ----------
+    hdrs: list of Header objects
+        The headers to be searched
+    key: str
+        The key to be searched for
```
Comment (sbillinge) [inline xpdan/db_utils.py]: The scans will be sorted by the values of this metadata key, e.g., 'sample_name'

File: xpdan/db_utils.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    verbose: bool, optional
+        If true prints the results. Defaults to True
+
+    Returns
+    -------
+    dict:
+        The dictionary of unique values and positions
```
Comment (sbillinge) [inline xpdan/db_utils.py]: The dictionary  containing {'key-value':[list of scan indices]}.  For example a search over 'sample_name' might return {'Ni':[0,1,2,3,4,9,10],'gold nanoparticles':[5,6,7,8]}

File: xpdan/db_utils.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    ----------
+    hdrs: list of Header objects
+        The headers to be diffed
+    verbose: bool, optional
+        If true prints the results. Defaults to True
+
+    Returns
+    -------
+    dict:
+        The dictionary of keys with different values across the scans.
```
Comment (sbillinge) [inline xpdan/db_utils.py]: with at least one different value across the scans.

**Comment by CJ-Wright** [comment]
@sbillinge ready for re-review

**Comment by sbillinge** [comment]
Let's go!

**Comment by CJ-Wright** [comment]
I'll take that as a ready to merge.

**Pull request merged**